### PR TITLE
Add EigenSym

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
Needs https://github.com/gonum/lapack/pull/108

**Comment by btracey** [comment]
PTAL @kortschak

File: mat64/eigen.go [hunk 1]
```diff
@@ -23,6 +29,90 @@ func symmetric(m *Dense) bool {
 	return true
 }
 
+// EigenSym is a type for creating and manipulating the Eigen decomposition of
```
Comment (vladimir-ch) [inline mat64/eigen.go]: s/Eigen decomposition/eigendecomposition/ ? The terminology does not seem to be completely unified but eigendecomposition is maybe a bit more common.

File: mat64/eigen.go [hunk 2]
```diff
@@ -23,6 +29,90 @@ func symmetric(m *Dense) bool {
...
+
+// Factorize computes the Eigen decomposition of the symmetric matrix a.
+// The Eigen decomposition is defined as
+//  A = P * D * P^-1
+// where D is a diagonal matrix containing the eigenvalues of the matrix, and
+// P is a matrix of the eigenvectors of A. If the vectors input argument is
+// false, the eigenvectors are not computed.
+//
+// Factorize returns whether the decomposition succeeded. If the decomposition
+// failed, routines that require a successful factorization will panic.
```
Comment (vladimir-ch) [inline mat64/eigen.go]: s/routines/methods/ ?

File: mat64/eigen.go [hunk 3]
```diff
@@ -23,6 +29,90 @@ func symmetric(m *Dense) bool {
...
+		s.vectors = nil
+		return false
+	}
+	s.vectorsComputed = vectors
+	s.values = w
+	s.vectors = NewDense(n, n, sd.mat.Data)
+	return true
+}
+
+// succFact returns true if the receiver contains a successful factorization.
```
Comment (vladimir-ch) [inline mat64/eigen.go]: s/true if/whether/

File: mat64/eigen.go [hunk 4]
```diff
@@ -23,6 +29,90 @@ func symmetric(m *Dense) bool {
...
+func (es *EigenSym) succFact() bool {
+	return len(es.values) != 0
+}
+
+// Values extracts the eigenvalues of the factorized matrix. If destination is
```
Comment (vladimir-ch) [inline mat64/eigen.go]: Please unify how dst is referred to in this paragraph. At the moment it is "destination", "receiver", and "dst".

File: mat64/eigen.go [hunk 5]
```diff
@@ -23,6 +29,90 @@ func symmetric(m *Dense) bool {
...
+// non-nil, the values are stored in-place into the receiver. In this case the
+// destination must have length n, otherwise Values will panic. If dst is
+// nil, then a new slice will be allocated of the proper length.
```
Comment (vladimir-ch) [inline mat64/eigen.go]: Say something about the returned slice?

File: mat64/eigen.go [hunk 6]
```diff
@@ -23,6 +29,90 @@ func symmetric(m *Dense) bool {
...
+	}
+	if len(dst) != len(es.values) {
+		panic(matrix.ErrSliceLengthMismatch)
+	}
+	copy(dst, es.values)
+	return dst
+}
+
+// EigenvectorsSym extracts the eigenvectors of the factorized matrix and stores
+// them in the reciever. Each eigenvector is a column of the matrix, with the
```
Comment (vladimir-ch) [inline mat64/eigen.go]: I don't like much the second part of this paragraph. How about:
```
... and stores them in the receiver whose columns are the eigenvectors corresponding to eigenvalues returned by Values.
```
or something like that?

File: mat64/eigen.go (hunk 1 above)
Comment (btracey) [inline mat64/eigen.go]: Changed to eigenvalue decomposition.

File: mat64/eigen.go (hunk 2 above)
Comment (btracey) [inline mat64/eigen.go]: Sure.

File: mat64/eigen.go (hunk 3 above)
Comment (btracey) [inline mat64/eigen.go]: Done.

File: mat64/eigen.go (hunk 5 above)
Comment (btracey) [inline mat64/eigen.go]: Done.

File: mat64/eigen.go (hunk 6 above)
Comment (btracey) [inline mat64/eigen.go]: Done (though modified)

**Comment by btracey** [comment]
PTAL

File: mat64/eigen.go [hunk 7]
```diff
@@ -23,6 +29,91 @@ func symmetric(m *Dense) bool {
...
+	if len(dst) != len(es.values) {
+		panic(matrix.ErrSliceLengthMismatch)
+	}
+	copy(dst, es.values)
+	return dst
+}
+
+// EigenvectorsSym extracts the eigenvectors of the factorized matrix and stores
+// them in the reciever. Each eigenvector is a column corresponding to the
+// respective eigenvalue returned by EigenSym.Values.
```
Comment (vladimir-ch) [inline mat64/eigen.go]: I think it should better say `es.Values` instead of `EigenSym.Values`.

**Comment by vladimir-ch** [comment]
LGTM

**Comment by kortschak** [comment]
Sorry, late to the party. The receiver should be consistent. I would suggest e throughout. I'll send a PR.

**Pull request merged**